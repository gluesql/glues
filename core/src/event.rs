use {
    crate::{
        data::{Directory, Note},
        types::DirectoryId,
    },
    strum_macros::Display,
};

#[derive(Clone, Debug, Display)]
pub enum Event {
    #[strum(to_string = "Key::{0}")]
    Key(KeyEvent),

    #[strum(to_string = "Entry::{0}")]
    Entry(EntryEvent),

    #[strum(to_string = "Notebook::{0}")]
    Notebook(NotebookEvent),

    Cancel,
}

#[derive(Clone, Debug, Display)]
pub enum EntryEvent {
    OpenMemory,
    OpenCsv(String),
    OpenJson(String),
    OpenFile(String),
    OpenGit {
        path: String,
        remote: String,
        branch: String,
    },
    OpenMongo {
        conn_str: String,
        db_name: String,
    },
}

#[derive(Clone, Debug, Display)]
pub enum NotebookEvent {
    OpenDirectory(DirectoryId),
    CloseDirectory(DirectoryId),

    SelectNote(Note),
    SelectDirectory(Directory),

    RenameNote(String),
    RenameDirectory(String),

    RemoveNote,
    RemoveDirectory,

    ShowNoteActionsDialog,
    CloseNoteActionsDialog,

    ShowDirectoryActionsDialog,
    CloseDirectoryActionsDialog,

    AddNote(String),
    AddDirectory(String),

    OpenNote,
    EditNote,
    ViewNote,
    BrowseNoteTree,

    UpdateNoteContent(String),

    CloseEntryDialog,
}

#[derive(Clone, Copy, Debug, Display)]
pub enum KeyEvent {
    B,
    E,
    H,
    J,
    K,
    L,
    M,
    O,
    Num(NumKey),
    Left,
    Right,
    Esc,
}

#[derive(Clone, Copy, Debug, Display)]
pub enum NumKey {
    One,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Zero,
}

impl From<EntryEvent> for Event {
    fn from(event: EntryEvent) -> Self {
        Self::Entry(event)
    }
}

impl From<NotebookEvent> for Event {
    fn from(event: NotebookEvent) -> Self {
        Self::Notebook(event)
    }
}

impl From<KeyEvent> for Event {
    fn from(event: KeyEvent) -> Self {
        Self::Key(event)
    }
}

impl From<NumKey> for KeyEvent {
    fn from(num_key: NumKey) -> Self {
        KeyEvent::Num(num_key)
    }
}

impl From<NumKey> for usize {
    fn from(num_key: NumKey) -> Self {
        match num_key {
            NumKey::One => 1,
            NumKey::Two => 2,
            NumKey::Three => 3,
            NumKey::Four => 4,
            NumKey::Five => 5,
            NumKey::Six => 6,
            NumKey::Seven => 7,
            NumKey::Eight => 8,
            NumKey::Nine => 9,
            NumKey::Zero => 0,
        }
    }
}

impl std::ops::Add<usize> for NumKey {
    type Output = usize;

    fn add(self, rhs: usize) -> Self::Output {
        let n = usize::from(self).saturating_add(rhs);

        if n > u16::MAX as usize {
            u16::MAX as usize
        } else {
            n
        }
    }
}
