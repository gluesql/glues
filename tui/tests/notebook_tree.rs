#[macro_use]
mod tester;
use tester::Tester;

use color_eyre::Result;
use ratatui::crossterm::event::KeyCode;

#[tokio::test]
async fn opens_note_on_l() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // select first note and open
    t.press('j').await;
    t.press('l').await;
    t.draw()?;

    // editor shows sample content
    snap!(t, "note_open");

    Ok(())
}

#[tokio::test]
async fn add_note_via_directory_actions() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // open directory actions on root and choose Add note
    t.press('m').await;
    t.draw()?;
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "add_note_prompt_open");

    // type new note name and submit
    for ch in "New Note".chars() {
        t.press(ch).await;
    }
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "note_added");

    Ok(())
}

#[tokio::test]
async fn add_directory_via_directory_actions() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // open directory actions on root and choose Add directory
    t.press('m').await;
    t.draw()?;
    t.press('j').await;
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "add_dir_prompt_open");

    for ch in "Tmp".chars() {
        t.press(ch).await;
    }
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "dir_added");

    Ok(())
}

#[tokio::test]
async fn rename_note_prompt_and_apply() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // select first note and open note actions: default selection is Rename note
    t.press('j').await;
    t.press('m').await;
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "rename_note_prompt_open");

    // clear default name and type new name
    for _ in 0..20 {
        t.key(KeyCode::Backspace).await;
    }
    for ch in "Renamed".chars() {
        t.press(ch).await;
    }
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "note_renamed");

    Ok(())
}

#[tokio::test]
async fn remove_note_confirm_cancel_then_accept() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // select note, open note actions, move to Remove note
    t.press('j').await;
    t.press('m').await;
    t.press('j').await;
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "remove_note_confirm");

    // cancel once
    t.press('n').await;
    t.draw()?;
    snap!(t, "remove_note_cancelled");

    // remove for real
    t.press('m').await;
    t.press('j').await;
    t.key(KeyCode::Enter).await;
    t.press('y').await;
    t.draw()?;
    snap!(t, "note_removed");

    Ok(())
}

#[tokio::test]
async fn remove_directory_confirm_cancel_then_accept() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // first add a directory so we can remove it
    t.press('m').await;
    t.draw()?;
    t.press('j').await; // Add directory
    t.key(KeyCode::Enter).await;
    for ch in "Tmp".chars() {
        t.press(ch).await;
    }
    t.key(KeyCode::Enter).await;
    t.draw()?;

    // select the new directory (first child under root)
    t.press('j').await;
    t.draw()?;

    // open directory actions and choose Remove directory
    t.press('m').await;
    t.press('j').await; // Add directory -> Rename directory
    t.press('j').await; // Rename directory -> Remove directory
    t.key(KeyCode::Enter).await;
    t.draw()?;
    snap!(t, "remove_dir_confirm");

    // cancel
    t.press('n').await;
    t.draw()?;
    snap!(t, "remove_dir_cancelled");

    // confirm remove
    t.press('m').await;
    t.press('j').await;
    t.press('j').await;
    t.key(KeyCode::Enter).await;
    t.press('y').await;
    t.draw()?;
    snap!(t, "dir_removed");

    Ok(())
}

#[tokio::test]
async fn note_actions_dialog_toggles() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // select note to enable note actions
    t.press('j').await;

    // open note actions dialog
    t.press('m').await;
    t.draw()?;
    snap!(t, "note_actions_open");

    // close dialog with Esc
    t.key(KeyCode::Esc).await;
    t.draw()?;
    snap!(t, "note_actions_closed");

    Ok(())
}

#[tokio::test]
async fn dir_actions_dialog_toggles() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // on root directory selection, open directory actions dialog
    t.press('m').await;
    t.draw()?;
    snap!(t, "dir_actions_open");

    // close dialog with Esc
    t.key(KeyCode::Esc).await;
    t.draw()?;
    snap!(t, "dir_actions_closed");

    Ok(())
}

 
