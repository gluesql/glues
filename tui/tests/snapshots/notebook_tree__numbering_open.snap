---
source: tui/tests/notebook_tree.rs
expression: text
snapshot_kind: text
---
 Steps: '2' selected                                                                                   [?] Show keymap 
[Browser]                                   ▐[Editor]                                                                   
 󰝰 Notes                                    ▐ 1 Welcome to Glues :D                                                     
   󱇗 Sample Note                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐
